agent/proxycfg/state.go

```diff
@@ -186,18 +187,43 @@ func (s *state) initWatchesConnectProxy() error {
 		case structs.UpstreamDestTypeService:
 			fallthrough
 		case "": // Treat unset as the default Service type
-			err = s.cache.Notify(s.ctx, cachetype.HealthServicesName, &structs.ServiceSpecificRequest{
-				Datacenter:   dc,
-				QueryOptions: structs.QueryOptions{Token: s.token},
-				ServiceName:  u.DestinationName,
-				Connect:      true,
-				// Note that Identifier doesn't type-prefix for service any more as it's
-				// the default and makes metrics and other things much cleaner. It's
-				// simpler for us if we have the type to make things unambiguous.
-			}, "upstream:"+serviceIDPrefix+u.Identifier(), s.ch)
 
-			if err != nil {
-				return err
+			// Determine if this should use a discovery chain.
+			//
+			// TODO(rb): reduce this list of exceptions
+			var shouldUseDiscoveryChain bool
+			if dc != s.source.Datacenter {
+				shouldUseDiscoveryChain = false
+			} else if u.DestinationNamespace != "" && u.DestinationNamespace != "default" {
+				shouldUseDiscoveryChain = false
```

Comment:
Hmmm. Why would cross-namespace requests not use the chain? I don't think it matters much for now because namespaces don't really exist yet but this is interesting...

Reply:
At the time I wrote this it was a combination of them not existing yet and also some other cache.Notify plumbing being irksome. I'll make a note to look at this again and see if i can delete the conditional.